### PR TITLE
[feature]Optimize the performance of Qwen2.5-VL on Ascend

### PR DESCRIPTION
**Pull request by ash-sigh**
## Motivation

Support w8a8-int8 quantization and optimize the performance of VIT for Qwen2.5-VL on Ascend

## Modifications

1. use `torch_npu.npu_mrope`, `torch_npu.npu_rotary_mul` and `torch_npu._npu_flash_attention_unpad` for NPU performance.
2. Padding the dimensions of hidden states from 80 to 128 with 0, which makes it more compatible with Ascend kernels.
3. Equivalently replace Conv3d in  **PatchEmbedding** with matmul to improve the overall performance.
4. run `image_processor` on CPU, because transformers has some limitations on Ascend.

## Accuracy Tests

launch_server :
Npu（w8a8-int）:
export HCCL_OP_EXPANSION_MODE="AIV"
export OMP_PROC_BIND=false
python -m sglang.launch_server --model-path /data/Qwen2.5-VL-72B-Insturct-W8A8-Int8 --tp 4 --host 127.0.0.1 --port 8022 --device npu --attention-backend ascend --quantization w8a8_int8

H20（fp8）:
python -m sglang.launch_server --model-path /data/Qwen2.5-VL-72B-Insturct-FP8-Static --tp 4 --host 127.0.0.1 --port 8022 --quantization fp8

benchmark mmmu:
python benchmark/mmmu/bench_sglang.py --port 8022 --concurrency 16 --dataset-path /data/MMMU

Result of H20:
Benchmark time: 284.83
Overall accuracy: 0.5

Result of Npu:
Benchmark time: 415.41
Overall accuracy: 0.609

## Benchmarking and Profiling

benchmark:
python3 -m sglang.bench_serving --backend sglang --host 127.0.0.1 --port 8022 --num-prompt 32 --dataset-name mmmu --max-concurrency 32 --random-output-len 1024 --apply-chat-template --flush-cache

result:
Mean TTFT(ms): 7463.16
Median TTFT(ms): 6864.70

Mean ITL(ms): 85.21
Median ITL(ms): 83.76

## Checklist
image

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/srt/models/qwen2_5_vl.py [hunk 1]
```diff
@@ -434,6 +596,178 @@ def forward(
...
+        )
+        rotary_pos_emb = rotary_pos_emb[window_index, :, :]
+        rotary_pos_emb = rotary_pos_emb.reshape(seq_len, -1)
+        position_embeddings = self.cal_cos_sin(rotary_pos_emb)
+
+        # compute cu_seqlens
+        cu_seqlens = (
+            torch.repeat_interleave(grid_thw[:, 1] * grid_thw[:, 2], grid_thw[:, 0])
+            .cpu()
+            .to(torch.int32)
```
[the remark this answers is not in the document]
Comment (ash-sigh) [reply python/sglang/srt/models/qwen2_5_vl.py]: By default, cu_window_seqlens and cu_seqlens run on CPU, there is no need to device-to-host transfer explicitly. Moveover, if the data run on NPU, it is prone to OOM.
Revision shown as below:
```python
window_index, cu_window_seqlens = self.get_window_index(grid_thw)
cu_window_seqlens = torch.tensor(
    cu_window_seqlens,
    dtype=torch.int32,
)
cu_window_seqlens = torch.unique_consecutive(cu_window_seqlens)
cu_window_seqlens = torch.diff(cu_window_seqlens)

seq_len, _ = x.size()

x = x.reshape(seq_len // self.spatial_merge_unit, self.spatial_merge_unit, -1)
x = x[window_index, :, :]
x = x.reshape(seq_len, -1)
rotary_pos_emb = rotary_pos_emb.reshape(
    seq_len // self.spatial_merge_unit, self.spatial_merge_unit, -1
)
rotary_pos_emb = rotary_pos_emb[window_index, :, :]
rotary_pos_emb = rotary_pos_emb.reshape(seq_len, -1)
position_embeddings = self.cal_cos_sin(rotary_pos_emb)

# compute cu_seqlens
cu_seqlens = (
    torch.repeat_interleave(grid_thw[:, 1] * grid_thw[:, 2], grid_thw[:, 0])
    .to(torch.int32)
)
```

File: python/sglang/srt/layers/rotary_embedding.py [hunk 2]
```diff
@@ -1076,6 +1077,34 @@ def forward(
...
+            mrope_section=mrope_section,
+            rotary_mode=rotary_mode,
+        )
+        return query, key
+
+    def mrope_dispatch_forward(self):
+        if _is_npu:
+            return self.forward_npu
+        else:
+            return self.forward
```
[the remark this answers is not in the document]
Comment (ash-sigh) [reply python/sglang/srt/layers/rotary_embedding.py]: Rename `MRotaryEmbedding.forward` to `MRotaryEmbedding.forward_native`

File: python/sglang/srt/multimodal/processors/base_processor.py [hunk 3]
```diff
@@ -231,6 +233,7 @@ def process_mm_data(
             hasattr(processor, "image_processor")
             and isinstance(processor.image_processor, BaseImageProcessorFast)
             and not self.server_args.disable_fast_image_processor
+            and not _is_npu
```
Comment (jgong5) [inline python/sglang/srt/multimodal/processors/base_processor.py:236]: Shall we call `is_npu()` directly instead of creating a new global variable in all the files? If you are concerned with the perf impact, you may use `lru_cache` on the `is_npu()` function.

File: python/sglang/srt/models/qwen2_5_vl.py
Comment (jgong5) [inline python/sglang/srt/models/qwen2_5_vl.py:1]: I feel mixing the device-specific modeling with the general one is not a good idea. Shall we implementation some registration mechanism so that each device backend can register their own modeling implementation and maintain their modeling in the device-specific folders? When the modeling is instantiated, the device-specific modeling can be instantiated according to the currently active accelerator. This might need an RFC though.

File: python/sglang/srt/models/qwen2_5_vl.py [hunk 4]
```diff
@@ -105,6 +114,137 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        assert k.dim() == 3, k.dim()
+        assert v.dim() == 3, v.dim()
+
+        # internvl
+        if self.qk_normalization:
+            q, k = self._apply_qk_norm(q, k)
+
+        output = torch.empty_like(q)
+        # operator requires pta version >= 2.5.1
+        torch_npu._npu_flash_attention_unpad(
```
Comment (mickqian) [inline python/sglang/srt/models/qwen2_5_vl.py:226]: We should probably introduce a new backend in `vision.py`

**Review by mickqian — CHANGES_REQUESTED** [review]

File: python/sglang/srt/multimodal/processors/base_processor.py (hunk 3 above)
Comment (ash-sigh) [reply python/sglang/srt/multimodal/processors/base_processor.py:236]: yes, use `is_npu()` instead of global variable `_is_npu`

**Comment by JustinTong0323** [comment]
The modeling file does not accept major logical changes unless they are performance optimizations. It must remain clear and readable. So plz not introduce platform specific codes in general modeling file.
